### ISSUE TITLE
Have 'func-name-matching' ignore defineProperty function values by default

### ISSUE THREAD

**Issue by Moeriki** [eslint/eslint#7423]
**What rule do you want to change?**

`func-name-matching`

**Does this change cause the rule to produce more or fewer warnings?**

Fewer.

**How will the change be implemented? (New option, new default behavior, etc.)?**

New default behavior where we allow `{ value }` property descriptor a different name than `value`.

**Please provide some example code that this change will affect:**

``` js
var model = {};

Object.defineProperty(model, 'key', {
  value: function key() {
    //
  }
});

console.log(model.key.name); // logs 'key'
```

**What does the rule currently do for this code?**

>  Function name `key` should match property name `value`

**What will the rule do after it's changed?**

No error.

**Comment by nzakas** [comment]
Yeah, this seems to be in direct violation of what the rule is trying to enforce. Seems like you should just use `eslint-disable` if you want an exception in that one location. I'm not sure there's a good argument for saying `Object.defineProperty()` is a universally-accepted special case.

**Comment by mysticatea** [comment]
Sounds reasonable to me.

```js
Object.defineProperty(model, 'foo', {
  value: function foo() {
    //
  },
  enumerable: true,
  configurable: true,
  writable: true
});
```

This is same as

```js
model.foo = function foo() {
    //
};
```

The function `foo` is the value of `model.foo`.
Also, `function value` in the stacktrace is not helpful to developers. `value` does not express what the function is. It should be `foo`.
So I think the rule should enforce the name to `foo` rather than `value`.

```js
Object.defineProperty(model, 'foo', {
    value: function value() {  //※ Function name `value` should match property name `foo`.
        //
    }
});

Object.defineProperty(model, {
    foo: {
        value: function value() {  //※ Function name `value` should match property name `foo`.
            //
        }
    }
});

Object.defineProperty(model, 'foo', {
    value: function foo() {  //※ OK
        //
    }
});

Object.defineProperty(model, {
    foo: {
        value: function foo() {  //※ OK
            //
        }
    }
});
```

**Comment by platinumazure** [comment]
@mysticatea Should this also be extended to `Object.defineProperties()` in your view? I imagine similar arguments could be made there.

**Comment by mysticatea** [comment]
@platinumazure Yes. And `Object.create` as well. I'm considering about the definition of getters/setters.

**Comment by Trott** [comment]
If it's a helpful data point, this came up in the Node.js project: https://github.com/nodejs/node/pull/10408/files/b417c9da84dfd122021a6c3c0e367ec9d6b67abc#r93858090

**Comment by platinumazure** [comment]
@Moeriki What do you think of [@mysticatea's proposal](https://github.com/eslint/eslint/issues/7423#issuecomment-255592836)? Is that close to what you're looking for?

It would be good if we could agree on a proposal and get that into the initial post, so the team has something concrete to evaluate.

**Comment by Moeriki** [comment]
The proposal suits my needs.

To be complete we could also include `Object.create`, as mentioned by @mysticatea.

```js
Object.create(Object.prototype, {
  foo: {
    value: function value() { //※ Function name `value` should match property name `foo`.
      //
    }
  }
});

Object.create(Object.prototype, {
  foo: {
    value: function foo() { //※ OK
      //
    }
  }
});
```

**Comment by nzakas** [comment]
I'm :-1: on changing default behavior, but ok with adding an option.

**Comment by not-an-aardvark** [comment]
After looking at this proposal again, I'm :+1: for adding an option.

**Comment by ljharb** [comment]
To clarify - so the option, when enabled, would exempt only object literals passed to `Object.defineProperty`, `Object.defineProperties`, and the second argument of `Object.create`?

What if I make a custom "define" function, to cover ES3 compatibility? How would I indicate to the rule that this function takes property descriptors as well?

**Comment by mysticatea** [comment]
I'm a champion for the option, maybe `{"addressDefineProperty": true}` or something like.

> To clarify - so the option, when enabled, would exempt only object literals passed to Object.defineProperty, Object.defineProperties, and the second argument of Object.create?

Yes, I think so.

> What if I make a custom "define" function, to cover ES3 compatibility? How would I indicate to the rule that this function takes property descriptors as well?

I think enough if it supports ES standard functions for now.

**Comment by platinumazure** [comment]
@ljharb Do you have an urgent need for that sort of support? If not, maybe
we could just cover the standard functions for now and we can consider
adding support for user-configured functions later if there is enough
demand.

**Comment by ljharb** [comment]
I do, in fact - I prefer "func-name-matching" set to "never", but many of my shim packages end up disabling that rule with inline comments every time i create a property descriptor with a function (whose name intentionally matches the key, in that special case). However, for compat, I pass things into a custom "define" function, and not into `Object.defineProperty` directly.

**Comment by mysticatea** [comment]
@ljharb Do you know major ES3 environments which are maintained officially?

**Comment by ljharb** [comment]
@mysticatea of course not, but "officially maintained" is utterly irrelevant; if users use it, a company/site is likely to choose to support it.

**Comment by platinumazure** [comment]
Thanks @ljharb, your point is well taken.

Let's step back a bit and just follow our process. If the team champions and endorses not only the standard property-descriptor functions, but also user-configurable ones, then that's what we'll implement. If the team champions and endorses just the standard property-descriptor functions, then we'll implement that instead and we can re-evaluate later if we get more demand for customizability. (And until that point, @ljharb can certainly create a custom rule for his use case.)

I'm :+1: for the standard property descriptor functions for sure-- it should be fairly easy to do. I'm not outright opposed to customizations, but I don't think we have enough demand from the community just yet.

**Comment by Moeriki** [comment]
I'd be willing to implement an option with changed behaviour.

Though I do not agree we should _exempt_ object literals passed in Object.defineProperty, etc… which seems to be the suggestion so far. I'd still enforce the name-match since we do know the registered name.

`{ considerPropertyDescriptor: true }`

> When set to true it will infer the function name from the property descriptor `name` instead of the direct object key.

**Comment by platinumazure** [comment]
I think between @mysticatea championing this (behind an option, for standard property descriptor related functions), my proposing the same later, and @nzakas and @not-an-aardvark also writing :+1: comments, we should be able to accept this. Does anyone disagree or object?

**Comment by not-an-aardvark** [comment]
Sounds fine to me.

**Comment by not-an-aardvark** [comment]
I went ahead and marked this as accepted based on @platinumazure's reasoning from https://github.com/eslint/eslint/issues/7423#issuecomment-321538608.

**Comment by kaicataldo** [comment]
Sorry to come into this so late in the game, but it's not entirely clear to me what the accepted implementation is here. Can we have a summary of the current proposed implementation and vote on it? The current :+1:'s don't seem to necessarily reflect support for the current proposal.

**Comment by platinumazure** [comment]
@kaicataldo Agreed, and I apologize for muddying the waters here.

@Moeriki Could you please either edit the initial post of this issue, or create a new issue, regarding the proposal that we should now be evaluating? The initial post mentions making this a default behavior change, but many on the team are against that. I think we've coalesced on creating a new option? Basically, could you please modify the issue to show what you're actually proposing to implement in the associated pull request? Thanks!

**Comment by platinumazure** [comment]
Closing this issue in favor of #9926, which has just been accepted.
